### PR TITLE
grandpa: implement accountable safety answers

### PR DESCRIPTION
**Pull request by marcio-diaz**
This builds on top of #2551 and implement the methods for `AccountableSafety`

### CODE REVIEW

File: core/finality-grandpa/src/aux_schema.rs [hunk 1]
```diff
@@ -425,17 +425,67 @@ pub(crate) fn update_authority_set<Block: BlockT, F, R>(
 	}
 }
 
-/// Write voter set state.
+/// Write voter set state and historical votes to database.
+pub(crate) fn write_voter_state<Block: BlockT, B: AuxStore>(
```
Comment (andresilva) [inline core/finality-grandpa/src/aux_schema.rs]: This name is a bit misleading especially wrt the method below.

File: core/finality-grandpa/src/aux_schema.rs [hunk 2]
```diff
@@ -632,4 +683,130 @@ mod test {
 			},
 		);
 	}
+
+	#[test]
+	fn load_decode_from_v2_migrates_data_format() {
```
Comment (andresilva) [inline core/finality-grandpa/src/aux_schema.rs]: Where is the migration code that's being tested?

File: core/finality-grandpa/src/environment.rs [hunk 3]
```diff
@@ -75,6 +74,7 @@ pub struct CompletedRounds<Block: BlockT> {
 	voters: Vec<AuthorityId>,
 }
 
+
```
Comment (andresilva) [inline core/finality-grandpa/src/environment.rs]:
```suggestion
```

File: core/finality-grandpa/src/aux_schema.rs [hunk 4]
```diff
@@ -425,17 +425,67 @@ pub(crate) fn update_authority_set<Block: BlockT, F, R>(
...
+	backend.insert_aux(
+		&[
+			(&votes_key[..], historical_votes.encode().as_slice()),
+			(SET_STATE_KEY, state.encode().as_slice()),
+		],
+		&[],
+	)
+}
+
+/// Write voter set state to database.
```
Comment (andresilva) [inline core/finality-grandpa/src/aux_schema.rs]: Is this still used?

File: core/finality-grandpa/src/aux_schema.rs [hunk 5]
```diff
@@ -425,17 +425,67 @@ pub(crate) fn update_authority_set<Block: BlockT, F, R>(
...
 	backend.insert_aux(
 		&[(SET_STATE_KEY, state.encode().as_slice())],
-		&[]
+		&[],
 	)
 }
 
+/// Write historical votes to database.
+#[cfg(test)]
+pub(crate) fn write_historical_votes<Block: BlockT, B: AuxStore>(
```
Comment (andresilva) [inline core/finality-grandpa/src/aux_schema.rs]: Is this used anywhere?

**Comment by gavofyork** [comment]
Looking a bit stale... is there a plan to get this in shape soon @marcio-diaz ?

**Comment by andresilva** [comment]
Needs to rebased as well, after #2551 is done.

**Comment by bkchr** [comment]
What is the status of this one? Just requiring a review?

**Comment by marcio-diaz** [comment]
> What is the status of this one? Just requiring a review?

It has a bunch of code not needed anymore for the same reason of the other PR.
I'll rewrite it in another PR, closing.

**Pull request closed without merging**